[WebDriver BiDi] Add foundation for script.evaluate implementation

sergedeh opened the pull request:
#### ffa4174157ba8a428bdb6a4d5e6beab1d15b546c
<pre>
[WebDriver BiDi] Add foundation for script.evaluate implementation

<a href="https://bugs.webkit.org/show_bug.cgi?id=288060">https://bugs.webkit.org/show_bug.cgi?id=288060</a>

Reviewed by BJ Burg.

This patch implements the foundational infrastructure for WebDriver BiDi
script.evaluate command. It establishes the protocol definitions,
IPC plumbing, and basic evaluation capability needed for subsequent
feature enhancements.

The implementation includes:
- Protocol definitions for script.evaluate in BidiScript.json
- IPC message routing from UIProcess to WebProcess
- JavaScript wrapper for BiDi-compliant script evaluation
- Basic RemoteValue serialization for primitive types
- Simplified exception handling with placeholder stack traces

This foundation supports basic script evaluation with context-based targets.
Advanced features (resultOwnership, serializationOptions, sandbox, realm
targets, complex types, detailed exception information) are marked with
FIXME comments for future implementation.

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
(WebKit::BidiScriptAgent::evaluate): Add basic script.evaluate implementation
with context-based target handling.
(WebKit::BidiScriptAgent::finishEvaluateBidiScriptResult): Add result
processing with simplified exception handling.
(WebKit::deserializeRemoteValue): Add basic primitive type deserialization.

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
Add method declarations for script.evaluate support.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::evaluateBidiScript): Add bridge method for
BiDi script evaluation.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
Add evaluateBidiScript method declaration.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::processBidiMessage): Simplify to rely on
backend dispatcher validation.

* Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
Update RemoteValueType enum with primitive types.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
Add IPC message handling for EvaluateBidiScript.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::evaluateBidiScript): Implement script
evaluation with BiDi envelope format.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
Add evaluateBidiScript method declaration.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:
(evaluateBidiScript): Add JavaScript wrapper for BiDi-compliant evaluation
with envelope format and type markers.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:
Add EvaluateBidiScript IPC message definition.

* WebDriverTests/TestExpectations.json:
Update test expectations for script.evaluate tests.

Canonical link: <a href="https://commits.webkit.org/306378@main">https://commits.webkit.org/306378@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9abf663e820a7561be2a0486570f4921ff827c79

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/140707 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/13089 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/2250 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/149058 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/93794 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/13801 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/13243 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/107903 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/78331 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/143658 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/10666 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/125967 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/88805 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/10275 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/7834 "Passed tests") | [⏳ wpe-libwebrtc ](None "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/119521 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/151670 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/12777 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/2137 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/116199 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/12792 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/11115 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/116537 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29738 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/12498 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/122593 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/67912 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/12819 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/2036 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/12559 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/76519 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/12758 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/12603 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -64,6 +67,89 @@ BidiScriptAgent::BidiScriptAgent(WebAutomationSession& session, BackendDispatche
+    // https://bugs.webkit.org/show_bug.cgi?id=288060
+    using namespace Inspector::Protocol::BidiScript;
+
+    auto out = RemoteValue::create();
```
Nit: `result` or `resultValue` is more clear IMO

burg reviewed (commented):
I only took a brief look at the PR so far, but this looks like very good progress!

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -64,6 +67,89 @@ BidiScriptAgent::BidiScriptAgent(WebAutomationSession& session, BackendDispatche
+                remoteValue->setValue(value.releaseNonNull());
+            return remoteValue;
+        }
+        // For any other unknown types, default to undefined
```
Nit: (applies throughout)

WebKit style is to use periods at the end of sentences in comments.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -64,6 +67,89 @@ BidiScriptAgent::BidiScriptAgent(WebAutomationSession& session, BackendDispatche
+{
+    // FIXME: Implement full BiDi RemoteValue deserialization (array, object, map, set, etc.)
+    // https://bugs.webkit.org/show_bug.cgi?id=288060
+    using namespace Inspector::Protocol::BidiScript;
```
Might be clearer to use

`using RemoteValueType = Inspector::Protocol::BidiScript::RemoteValueType`

etc similar to how it's done in Python. Makes it easier to follow the definitions.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -126,37 +212,148 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+        return;
+    }
+
+    // Success case
```
Nit: comment unnecessary

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -126,37 +212,148 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
-        if (result)
-            resultObject->setValue(JSON::Value::create(WTF::move(result.value())));
+    auto envelopePayload = JSON::Value::parseJSON(result.value());
+    auto envelopeObj = envelopePayload ? envelopePayload->asObject() : nullptr;
```
Nit: `envelopeObject`

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -126,37 +212,148 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
-    });
+    // Handle malformed envelope (internal error - failed to parse or not an object)
+    if (!envelopeObj) {
+        String errorText = makeString("Internal error: malformed BiDi envelope (failed to parse result as JSON object)"_s);
```
Suggested message:

Internal error: malformed envelope in call to `script.evaluate`

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -126,37 +212,148 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+    // Handle malformed envelope (internal error - failed to parse or not an object)
+    if (!envelopeObj) {
+        String errorText = makeString("Internal error: malformed BiDi envelope (failed to parse result as JSON object)"_s);
+        auto exceptionRemote = BidiScript::RemoteValue::create()
```
I'm not quite sure what this argument is for.

sergedeh replied:
exceptionRemote  is the exception field (a script.RemoteValue object required by the BiDi spec)   (https://w3c.github.io/webdriver-bidi/#type-script-ExceptionDetails). It represents the exception value itself in the protocol format.
I've updated the code to properly set a value map on the RemoteValue (with at least a message field) as the spec requires, making it a proper map object rather than just a type-only RemoteValue.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
```diff
@@ -78,6 +80,16 @@ class BidiScriptAgent final : public Inspector::BidiScriptBackendDispatcherHandl
+    void finishEvaluateBidiScriptResult(const String& realmId, const String& expression, Inspector::CommandResult<String>&&, Inspector::CommandCallbackOf<Inspector::Protocol::BidiScript::EvaluateResultType, String, RefPtr<Inspector::Protocol::BidiScript::RemoteValue>, RefPtr<Inspector::Protocol::BidiScript::ExceptionDetails>>&&);
+
+    // Stubbed realm registry to decouple evaluate from fabricated ids.
+    class RealmRegistryStub {
```
Its kind of weird to use this inline class definition for something that probably will become its own class later. Can you put `StubRealmRegistry` class interface at the top level in this file, or within BidiScriptAgent as a nested class?

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
```diff
@@ -69,6 +70,7 @@ class BidiScriptAgent final : public Inspector::BidiScriptBackendDispatcherHandl
     void getRealms(const Inspector::Protocol::BidiBrowsingContext::BrowsingContext& optionalBrowsingContext , std::optional<Inspector::Protocol::BidiScript::RealmType>&& optionalRealmType, Inspector::CommandCallback<Ref<JSON::ArrayOf<Inspector::Protocol::BidiScript::RealmInfo>>>&&) override;
 
 private:
+    // getRealms implementation methods
```
Nit: comment doesn't add much

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -126,37 +212,148 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
+}
+
+// RealmRegistryStub implementation
+String BidiScriptAgent::RealmRegistryStub::realmIdForContext(const String& contextId) const
```
See comment in header file.

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:
```diff
@@ -123,6 +130,57 @@ let AutomationSessionProxy = class AutomationSessionProxy
+        if (callbackTimeout >= 0) {
+            timeoutPromise = new Promise((resolve, reject) => {
+                timeoutIdentifier = setTimeout(() => {
+                    reject({ name: "JavaScriptTimeout", message: "script timed out after " + callbackTimeout + "ms" });
```
Nit: use template string

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:
```diff
@@ -123,6 +130,57 @@ let AutomationSessionProxy = class AutomationSessionProxy
+                    // Handle promises for awaitPromise: true
+                    result.then(
+                        value => {
+                            let serialized = this.serializeBidiRemoteValue(value, maxObjectDepth);
```
Nit: `serializedValue`

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:
```diff
@@ -123,6 +130,57 @@ let AutomationSessionProxy = class AutomationSessionProxy
+                let __src = expression;
+                let result = globalThis.eval(__src);
+
+                if (awaitPromise && result && typeof result.then === 'function') {
```
Nit: use "

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:
```diff
@@ -123,6 +130,57 @@ let AutomationSessionProxy = class AutomationSessionProxy
+        let promise = new Promise((resolve, reject) => {
+            try {
+                // Execute expression using globalThis.eval pattern (like original implementation)
+                let __src = expression;
```
We don't really use `__` in this way. Is there a reason to name it so?

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:
```diff
@@ -123,6 +130,57 @@ let AutomationSessionProxy = class AutomationSessionProxy
+                            resolve({ __ok: true, remote: serialized });
+                        },
+                        error => {
+                            reject({ __ok: false, e: { name: error.name, message: error.message, stack: error.stack } });
```
Overall this code looks good, but `__ok` and `e` and `remote` could use more descriptive names. There isn't a runtime penalty for using clearer key names in the envelope object.

```
resolve({success: true, result: serializedValue})
resolve({success: false, error: {...}})

```

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:
```diff
@@ -123,6 +130,57 @@ let AutomationSessionProxy = class AutomationSessionProxy
+            }
+        });
+
+        // Handle timeout race
```
Nit: comment unnecessary

burg commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:
```diff
@@ -271,6 +329,114 @@ let AutomationSessionProxy = class AutomationSessionProxy
+            }
+
+            // Handle specialized object types
+            if (value instanceof Date) {
```
I'm OK with bending style rules for long section like this, but let's be consistent. Let's stick to single-line `if` without braces, and having the `return` statement be indented. Keep multiline if's as-is.

burg reviewed (approved):
Very nice, I mostly have style comments at this point.

burg reviewed (approved):
Excellent work!

sergedeh committed:
[WebDriver BiDi] Add foundation for script.evaluate implementation

https://bugs.webkit.org/show_bug.cgi?id=288060

Reviewed by BJ Burg.

This patch implements the foundational infrastructure for WebDriver BiDi
script.evaluate command. It establishes the protocol definitions,
IPC plumbing, and basic evaluation capability needed for subsequent
feature enhancements.

The implementation includes:
- Protocol definitions for script.evaluate in BidiScript.json
- IPC message routing from UIProcess to WebProcess
- JavaScript wrapper for BiDi-compliant script evaluation
- Basic RemoteValue serialization for primitive types
- Simplified exception handling with placeholder stack traces

This foundation supports basic script evaluation with context-based targets.
Advanced features (resultOwnership, serializationOptions, sandbox, realm
targets, complex types, detailed exception information) are marked with
FIXME comments for future implementation.

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
(WebKit::BidiScriptAgent::evaluate): Add basic script.evaluate implementation
with context-based target handling.
(WebKit::BidiScriptAgent::finishEvaluateBidiScriptResult): Add result
processing with simplified exception handling.
(WebKit::deserializeRemoteValue): Add basic primitive type deserialization.

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
Add method declarations for script.evaluate support.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::evaluateBidiScript): Add bridge method for
BiDi script evaluation.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
Add evaluateBidiScript method declaration.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::processBidiMessage): Simplify to rely on
backend dispatcher validation.

* Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
Update RemoteValueType enum with primitive types.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
Add IPC message handling for EvaluateBidiScript.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::evaluateBidiScript): Implement script
evaluation with BiDi envelope format.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
Add evaluateBidiScript method declaration.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.js:
(evaluateBidiScript): Add JavaScript wrapper for BiDi-compliant evaluation
with envelope format and type markers.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:
Add EvaluateBidiScript IPC message definition.

* WebDriverTests/TestExpectations.json:
Update test expectations for script.evaluate tests.

Canonical link: https://commits.webkit.org/306378@main
